Modules/Core/Common/test/itkImageDuplicatorTest2.cxx

```diff
@@ -69,9 +67,7 @@ itkImageDuplicatorTest2(int argc, char * argv[])
     dup->Update();
     ImageType::ConstPointer dupImage = dup->GetOutput();
 
-    writer->SetInput(dupImage);
-    writer->SetFileName(argv[2]);
-    writer->Update();
+    itk::ImageFileWriter<>::WriteImage(*dupImage, argv[2]);
```

Comment:
I don't like having to write `*varName`, I prefer just `varName`. See declaration suggestion on how to accomplish this.

Reply:
My proposal was to use an image reference, rather than a pointer, because a pointer in C++ is used to indicate "might be null". So I wonder, why would someone want to call `WriteImage` with a (possible) null pointer as image?

A smart pointer is specifically used to indicate that the ownership on an object is shared between entities. It does make sense to share with an ("old style") `ImageFileWriter<ImageType>` object, but I wonder if it makes sense to share with a static member function call.

That's why I'm reluctant to declare the image parameter as a pointer. But I do see your point that it is cumbersome to always add a `*`.

Reply:
@N-Dekker If a user has pointer which might be null, and wants to print it if it is not null, he will have to write an `if`. And if we require a reference, he will have to write an if before converting it to a reference. As all ITK images are held through pointers or smart pointers, accepting those is more natural than expecting a reference. I don't see much practical difference between the two, except that we will always have to add `*` during invocation with the reference approach.

Reply:
@dzenanz Just a few links to the C++ Core Guidelines to clarify my reluctance to declare the image as a pointer (smart or raw pointer):

"Take smart pointers as parameters only to explicitly express lifetime semantics"
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rr-smartptrparam

"For general use, take T* or T& arguments rather than smart pointers"
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers

"For “in” parameters, pass cheaply-copied types by value and others by reference to const"
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers

"Prefer T* over T& when “no argument” is a valid option"
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#f60-prefer-t-over-t-when-no-argument-is-a-valid-option 

Now I know that there are a lot of old ITK (member) functions that have pointers as argument types. But for a newly added function maybe we could move forward to use a reference as argument type instead...

Reply:
We would be explicitly taking `itk::SmartPointer<Image<...>>` in order to aid template parameter deduction. And primary specialization would be taking `Image *`. And if you want to be super-correct, you can check for null to avoid crashing.

As we can only hold a (smart) pointer to an ITK image in a variable, this is what we should be passing around. Entire ITK API takes either `Image *` or `Image::Pointer` (aka `itk::SmartPointer<Image<...>>`). Trying to switch one function to taking an `Image &` seems unnecessary and would make its use harder (specifically having to dereference image at each place of invocation).

Reply:
And as you might guess, refactoring ITK to change convention from passing `Image*` to passing `Image&` is a huge task which breaks backwards compatibility with little tangible gain.

Reply:
> And if you want to be super-correct, you can check for null to avoid crashing.

This function would likely be used for debugging, so it would be good add a null check and a check that the image has been allocated.

There are other issues with the wrapping to use raw pointers instead of references.